packages/adapter-utils/src/utils.ts

```diff
@@ -101,33 +125,112 @@ const fieldMapperGenerator = (
   )
 }
 
-export type TransformFuncArgs = {
-  value: Value
-  path?: ElemID
-  field?: Field
+const removeEmptyParts = (value: Value, allowEmpty: boolean): Value => {
+  if (Array.isArray(value)) {
+    const filtered = value.filter(isDefined)
+    return filtered.length === 0 && (value.length > 0 || !allowEmpty) ? undefined : filtered
+  }
+  if (_.isPlainObject(value)) {
+    const filtered = _.omitBy(value, _.isUndefined)
+    return _.isEmpty(filtered) && (!_.isEmpty(value) || !allowEmpty) ? undefined : filtered
+  }
+  return value
 }
-export type TransformFunc = (args: TransformFuncArgs) => Promise<Value> | Value | undefined
 
-export type TransformFuncSync = (args: TransformFuncArgs) => lowerDashTypes.NonPromise<Value> | undefined
+const recurseIntoValue = (
+  newVal: Value,
+  transformFunc: (value: Value, keyPathID?: ElemID, field?: Field) => Value,
+  strict: boolean,
+  allowEmpty: boolean,
+  isAsync: boolean,
+  keyPathID?: ElemID,
+  field?: Field,
+  fieldType?: TypeElement,
+): Value => {
+  if (newVal === undefined) {
+    return undefined
+  }
 
-type TransformValuesBaseArgs = {
-  values: Value
-  type: ObjectType | TypeMap | MapType | ListType
-  strict?: boolean
-  pathID?: ElemID
-  isTopLevel?: boolean
-  allowEmpty?: boolean
+  if (isReferenceExpression(newVal)) {
+    return newVal
+  }
+
+  const listMapFunc = isAsync
+    ? (list: Iterable<Value>,
+      mapFunc: (t: Value, index: number) => Promise<unknown>) => toArrayAsync(mapAsync(list, mapFunc))
+    : _.map
+  const objMapFunc = isAsync
+    ? (obj: Iterable<Value>, mapFunc: (val: Value, key: string) => Promise<unknown>) => mapValuesAsync(obj, mapFunc)
+    : _.mapValues
```

Comment:
1. I think this would be easier to read with a helper type
2. the input for map values is an object, not an iterable
```suggestion
  type AsyncMapFunc<K> = (t: Value, key: K) => Promise<unknown>
  const listMapFunc = isAsync
    ? (list: Iterable<Value>, mapFunc: AsyncMapFunc<number>) => toArrayAsync(mapAsync(list, mapFunc))
    : _.map
  const objMapFunc = isAsync
    ? (obj: Record<string, Value>, mapFunc: AsyncMapFunc<string>) => mapValuesAsync(obj, mapFunc)
    : _.mapValues
```